core/src/validator.rs

```diff
@@ -1,256 +1,212 @@
 //! Structures and impls related to *runtime* `Validator`s processing.
 
 use core::fmt::{self, Debug, Formatter};
+use std::sync::RwLock;
 
-use dashmap::DashMap;
 use iroha_data_model::{
-    permission::validator::{
-        DenialReason, Id, NeedsPermission as _, NeedsPermissionBox, Type, Validator,
-    },
-    prelude::Account,
-    Identifiable,
+    permission::validator as data_model_validator, prelude::Account, Identifiable,
 };
+#[cfg(not(test))]
 use iroha_logger::trace;
-use iroha_primitives::must_use::MustUse;
 
 use super::wsv::WorldStateView;
 use crate::smartcontracts::wasm;
 
 /// [`Chain`] error type.
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
+    /// [`validate()`](Validator::validate) has been called before
+    /// initializing [`Validator`] with [`update()`](Validator::update).
+    #[error("Attempt to use uninitialized validator")]
+    Uninitialized,
     /// [`wasm`] module error.
     #[error("WASM error: {0}")]
     Wasm(#[from] wasm::Error),
     /// Validator denied the operation.
-    #[error("Validator `{validator_id}` denied the operation `{operation}`: `{reason}`")]
+    #[error("Validator denied the operation `{operation}`: `{reason}`")]
     ValidatorDeny {
-        /// Validator ID.
-        validator_id: <Validator as Identifiable>::Id,
         /// Denial reason.
-        reason: DenialReason,
+        reason: data_model_validator::DenialReason,
         /// Denied operation.
-        operation: NeedsPermissionBox,
+        operation: data_model_validator::NeedsPermissionBox,
     },
 }
 
-/// Result type for [`Chain`] operations.
+/// Result type for [`Validator`] operations.
 pub type Result<T, E = Error> = core::result::Result<T, E>;
 
-/// Chain of *runtime* permission validators. Used to validate operations that require permissions.
+/// Validator that checks if user has permission to perform some operation.
 ///
-/// Works similarly to the
-/// [`Chain of responsibility`](https://en.wikipedia.org/wiki/Chain-of-responsibility_pattern).
-/// The validation of an operation is forwarded to all
-/// validators in the chain which have the required type.
-/// The validation stops at the first
-/// [`Deny`](iroha_data_model::permission::validator::Verdict::Deny) verdict.
-#[derive(Clone)]
-pub struct Chain {
-    all_validators: DashMap<Id, LoadedValidator>,
-    concrete_type_validators: DashMap<Type, Vec<Id>>,
+/// Can be updated with instruction.
+pub struct Validator {
+    /// Optional pre-loaded validator. `None` when creating with [`Default::default()`].
+    /// Can be set with [`update()`](Validator::update).
+    loaded_validator: RwLock<Option<LoadedValidator>>,
```

Comment:
```suggestion
    loaded_validator: RwLock<LoadedValidator>,
```
this should not be optional. What should be optional is `World::validator`

Reply:
actually even the `RwLock` should be moved into `World::validator` IMO

Reply:
Ok, will do

Reply:
After trying I came out with the following results:

1. It's not convinient, because other fields of `World` has interior mutability
2. With that it will be unreal to have `validator_view()` or something like that in `WorldStateView`. Instead we should provide `validate()` directly. And that breaks the idea I had long before this PR.

My idea is that `WorldStateView` shouldn't copy-pase methods from `World` fields. It should provide a constant view to that fields. With that we could significantly reduce the amount of methods in this *God Object*. And my approach with `validator_view()` is a first step to it.

Reply:
at least you should move the option, no? Your comment is only about the `RwLock`?

Reply:
Hm, yes, I'll try to move `Option`